modules/reactor/src/meshdivisions/HexagonalGridDivision.C

```diff
@@ -0,0 +1,194 @@
+
+  if (!_center_positions)
+    setNumDivisions(_hex_latt->totalPins(_nr) * _nz);
+  else
+    setNumDivisions(_center_positions->getNumPositions() * _hex_latt->totalPins(_nr) * _nz);
+
+  // Check that the grid is well-defined
+  if (_center_positions)
+  {
+    Real min_center_dist = _center_positions->getMinDistanceBetweenPositions();
+    // Note that if the positions are not aligned on a hexagonal lattice themselves,
+    // this bound is not sufficiently strict. The simplest example would be non-coplanar
+    // points, which can be a great distance away axially but be on the same axis
+    if (MooseUtils::absoluteFuzzyGreaterThan(_lattice_flat_to_flat, min_center_dist))
+      mooseError(
+          "Hexagonal grids centered on the positions are too close to each other (min distance: ",
+          min_center_dist,
+          "), closer than the extent of each grid (",
+          _lattice_flat_to_flat,
+          "). Mesh division is ill - defined ");
+  }
+}
+
+unsigned int
+HexagonalGridDivision::divisionIndex(const Elem & elem) const
+{
+  return divisionIndex(elem.vertex_average());
+}
+
+unsigned int
+HexagonalGridDivision::divisionIndex(const Point & pt) const
+{
+  unsigned int offset = 0;
+
+  // Get point in the coordinates of the lattice. This can involve a projection due to
+  // the axis of the lattice, or simply a translation if there are lattices distributed
+  // using positions
+  Point pc;
+  if (_center_positions)
+  {
+    // If dividing using positions, find the closest position and
+    // look at the relative position of the point compared to that position
+    const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
+    const auto nearest_grid_center_index = _center_positions->getNearestPositionIndex(pt, initial);
+    offset = nearest_grid_center_index * _hex_latt->totalPins(_nr) * _nz;
+    const auto nearest_grid_center =
+        _center_positions->getPosition(nearest_grid_center_index, initial);
+
+    // Project in local hexagonal grid
+    pc = pt - nearest_grid_center;
+  }
+  else
+    pc = pt - _center;
+
+  // Get radial division index, using the channel as the pins are 0-radius
+  // The logic in get pin index requires getting the point in the plane of the pin centers
+  auto ir = _hex_latt->pinIndex(pc);
+  const auto n_pins = _hex_latt->nPins();
+
+  if (!_outside_grid_counts_as_border)
+  {
+    if (ir == n_pins)
+      return MooseMeshDivision::INVALID_DIVISION_INDEX;
+    if (MooseUtils::absoluteFuzzyLessThan(pc(_z_axis_index), _min_z) ||
+        MooseUtils::absoluteFuzzyGreaterThan(pc(_z_axis_index), _max_z))
+      return MooseMeshDivision::INVALID_DIVISION_INDEX;
+  }
+
+  // If too far from the grid to have a valid radial index, use the closest pin
+  if (ir == n_pins)
+    ir = _hex_latt->closestPinIndex(pc);
+
+  // Find axial index
+  const auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
```

Comment:
```suggestion
  constexpr auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
```

Reply:
not compiling
```
/Users/c/projects/moose_v1/modules/reactor/src/meshdivisions/HexagonalGridDivision.C:158:18: error: constexpr variable 'not_found' must be initialized by a constant expression
  constexpr auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
                 ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/c/projects/moose_v1/modules/reactor/src/meshdivisions/HexagonalGridDivision.C:158:30: note: read of non-const variable 'INVALID_DIVISION_INDEX' is not allowed in a constant expression
  constexpr auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
```

Reply:
Can you make `MooseMeshDivision::INVALID_DIVISION_INDEX` a constant expression?

Reply:
I think the way forward will likely one day be to have the outside index to be a user settable parameter.
So I d rather not change that for now